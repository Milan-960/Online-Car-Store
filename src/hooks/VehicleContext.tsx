import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import { Vehicle } from "../custom-types/Index";

interface VehicleContextProps {
  vehicles: Vehicle[];
  scrollPosition: number;
  setVehicles: Dispatch<SetStateAction<Vehicle[]>>;
  setScrollPosition: Dispatch<SetStateAction<number>>;
  modelFilter: string;
  categoryFilter: string;
  setModelFilter: Dispatch<SetStateAction<string>>;
  setCategoryFilter: Dispatch<SetStateAction<string>>;
  resetScrollPosition: () => void;
}

export const VehicleContext = createContext<VehicleContextProps>({
  vehicles: [],
  scrollPosition: 0,
  setVehicles: () => {},
  setScrollPosition: () => {},
  modelFilter: "",
  categoryFilter: "",
  setModelFilter: () => {},
  setCategoryFilter: () => {},
  resetScrollPosition: () => {},
});

interface VehicleProviderProps {
  children: ReactNode;
}

export const VehicleProvider: React.FC<VehicleProviderProps> = ({
  children,
}) => {
  const [vehicles, setVehicles] = useState<Vehicle[]>([]);
  const [scrollPosition, setScrollPosition] = useState<number>(0);
  const [modelFilter, setModelFilter] = useState<string>("");
  const [categoryFilter, setCategoryFilter] = useState<string>("");

  const resetScrollPosition = () => setScrollPosition(0);

  useEffect(() => {
    window.scrollTo(0, scrollPosition);
  }, [scrollPosition]);

  return (
    <VehicleContext.Provider
      value={{
        vehicles,
        setVehicles,
        scrollPosition,
        setScrollPosition,
        modelFilter,
        setModelFilter,
        categoryFilter,
        setCategoryFilter,
        resetScrollPosition,
      }}
    >
      {children}
    </VehicleContext.Provider>
  );
};
