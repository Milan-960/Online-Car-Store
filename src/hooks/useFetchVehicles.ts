import { useState, useEffect, useCallback, useContext } from "react";
import { Vehicle } from "../custom-types/Index";
import { getVehicles } from "../api/api";
import { VehicleContext } from "./VehicleContext";

//* PS: This compoent has a lot of filter logic however there are few ways
//* we can refactore the code creating Interface and just one handleFilterChange with all of the filter iterte thought that function
//* and set the state based on the filter name and value. and pass it to the ListingPage filter
//* I was running out of time so I was not able to refactore this :(

const useFetchVehicles = () => {
  const [loading, setLoading] = useState(true);
  const [vehicles, setVehicles] = useState<Vehicle[]>([]); // All vehicles data
  const [error, setError] = useState(null);
  const { setScrollPosition } = useContext(VehicleContext);

  // Aggregated data for filter values
  const [models, setModels] = useState<string[]>([]);
  const [categories, setCategories] = useState<string[]>([]);
  const [fuelTypes, setFuelTypes] = useState<string[]>([]);
  const [climateTypes, setClimateTypes] = useState<string[]>([]);
  const [interiorColors, setInteriorColors] = useState<string[]>([]);

  // Filter values for each filter field
  const [colorFilter, setColorFilter] = useState("");
  const [modelFilter, setModelFilter] = useState("");
  const [categoryFilter, setCategoryFilter] = useState("");
  const [fuelTypeFilter, setFuelTypeFilter] = useState("");
  const [climateFilter, setClimateFilter] = useState("");
  const [minPriceFilter, setMinPriceFilter] = useState(3000);
  const [maxPriceFilter, setMaxPriceFilter] = useState(100000);
  const [minRegYearFilter, setMinRegYearFilter] = useState(2010);
  const [maxRegYearFilter, setMaxRegYearFilter] = useState(2030);
  const [minKmFilter, setMinKmFilter] = useState(5000);
  const [maxKmFilter, setMaxKmFilter] = useState(100000);
  const [minEnginePowerFilter, setMinEnginePowerFilter] = useState(20);
  const [maxEnginePowerFilter, setMaxEnginePowerFilter] = useState(300);

  const [filterApplied, setFilterApplied] = useState(false);

  const [hasMore, setHasMore] = useState(true); // extracting the data state to determine if more data is available to fetch

  // Function to fetch vehicle data, with optional pagination
  const fetchVehicles = useCallback(
    async (page: number = 1, perPage: number = 6) => {
      setLoading(true);
      setError(null);

      // Fetch data from API
      try {
        const response = await getVehicles(page);

        console.log("res", response);

        // If there are records in the response, set hasMore to true
        setHasMore(response.records.length > 0);

        // If the response contains aggregated data, set the state variables
        if (
          response.aggregations?.model &&
          response.aggregations?.category &&
          response.aggregations?.["drivetrain.fuel.type"] &&
          response.aggregations?.climate &&
          response.aggregations?.["interior.interior_color"]
        ) {
          setModels(Object.keys(response.aggregations.model));
          setCategories(Object.keys(response.aggregations.category));
          setFuelTypes(
            Object.keys((response.aggregations as any)["drivetrain.fuel.type"])
          );
          setClimateTypes(Object.keys(response.aggregations.climate));
          setInteriorColors(
            Object.keys(response.aggregations["interior.interior_color"])
          );
        }

        // Apply filters to the fetched vehicles if any filter is selected
        let newVehicles = response.records;

        // Filter by model
        if (modelFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.model.toLowerCase() === modelFilter.toLowerCase()
          );
        }

        // Filter by category
        if (categoryFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.category.toLowerCase() === categoryFilter.toLowerCase()
          );
        }

        // Filter by fuel type
        if (fuelTypeFilter) {
          newVehicles = newVehicles.filter(
            ({
              drivetrain: {
                fuel: { type },
              },
            }: Vehicle) => type.toLowerCase() === fuelTypeFilter.toLowerCase()
          );
        }

        // Filter by climate
        if (climateFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.climate.toLowerCase() === climateFilter.toLowerCase()
          );
        }

        // Filter by color
        if (colorFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.color?.toLowerCase() === colorFilter.toLowerCase()
          );
        }

        // Filter by price range
        if (minPriceFilter || maxPriceFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.price_data.price >= minPriceFilter &&
              vehicle.price_data.price <= maxPriceFilter
          );
        }

        // Filter by registration year range
        if (minRegYearFilter || maxRegYearFilter) {
          newVehicles = newVehicles.filter((vehicle) => {
            if (vehicle.vehicle_history && vehicle.vehicle_history.reg_date) {
              const regYear = Number(
                vehicle.vehicle_history.reg_date.split(".")[1]
              );
              return regYear >= minRegYearFilter && regYear <= maxRegYearFilter;
            }
            return true; // Skip if not available
          });
        }

        // Filter by kilometer range
        if (minKmFilter || maxKmFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              vehicle.kilometer >= minKmFilter &&
              vehicle.kilometer <= maxKmFilter
          );
        }

        // Filter by engine power range
        if (minEnginePowerFilter || maxEnginePowerFilter) {
          newVehicles = newVehicles.filter(
            (vehicle) =>
              (vehicle.technical_features?.engine_power_max || Infinity) >=
                minEnginePowerFilter &&
              (vehicle.technical_features?.engine_power_max || 0) <=
                maxEnginePowerFilter
          );
        }

        // Add the newly fetched and filtered vehicles to the existing list
        setVehicles((prevVehicles) => [...prevVehicles, ...newVehicles]);
      } catch (error: any) {
        setError(error.message);
      } finally {
        setLoading(false);
      }
    },
    // The fetchVehicles function will re-run whenever any filter value changes
    [
      modelFilter,
      categoryFilter,
      fuelTypeFilter,
      climateFilter,
      colorFilter,
      minPriceFilter,
      maxPriceFilter,
      minRegYearFilter,
      maxRegYearFilter,
      minKmFilter,
      maxKmFilter,
      minEnginePowerFilter,
      maxEnginePowerFilter,
    ]
  );

  const loadMoreVehicles = () => {
    // Pass the number of items to fetch in the loadMoreVehicles function
    const nextPage = vehicles.length / 6 + 1;
    fetchVehicles(nextPage, 6);
  };

  const resetScrollPosition = () => {
    if (typeof window !== "undefined") {
      window.scrollTo(0, 0);
    }
  };

  // After each filter change function
  const handleModelChange = (model: string) => {
    setModelFilter(model);
    setFilterApplied(true);
    resetScrollPosition();
  };

  const resetFilters = () => {
    setFilterApplied(false);
    resetScrollPosition();
  };

  const handleCategoryChange = (category: string) => {
    setCategoryFilter(category);
    setFilterApplied(true);
    setScrollPosition(0);
  };

  const handleFuelTypeChange = (fuelType: string) => {
    setFuelTypeFilter(fuelType);
    setScrollPosition(0);
  };

  const handleClimateChange = (climate: string) => {
    setClimateFilter(climate);
    setScrollPosition(0);
  };

  const handleInteriorColorChange = (color: string) => {
    setColorFilter(color);
    setScrollPosition(0);
  };

  const handleMinPriceChange = (price: number) => {
    setMinPriceFilter(price);
    setScrollPosition(0);
  };

  const handleMaxPriceChange = (price: number) => {
    setMaxPriceFilter(price);
    setScrollPosition(0);
  };

  const handleMinRegYearChange = (regYear: number) => {
    setMinRegYearFilter(regYear);
    setScrollPosition(0);
  };

  const handleMaxRegYearChange = (regYear: number) => {
    setMaxRegYearFilter(regYear);
    setScrollPosition(0);
  };

  const handleMinKmChange = (km: number) => {
    setMinKmFilter(km);
    setScrollPosition(0);
  };

  const handleMaxKmChange = (km: number) => {
    setMaxKmFilter(km);
    setScrollPosition(0);
  };

  const handleMinEnginePowerChange = (enginePower: number) => {
    setMinEnginePowerFilter(enginePower);
    setScrollPosition(0);
  };

  const handleMaxEnginePowerChange = (enginePower: number) => {
    setMaxEnginePowerFilter(enginePower);
    setScrollPosition(0);
  };

  // Fetch data initially and whenever filterApplied state changes
  useEffect(() => {
    setVehicles([]);
    fetchVehicles();
  }, [fetchVehicles, filterApplied]);

  // Clear the current vehicles and re-fetch whenever any filter value changes
  useEffect(() => {
    if (filterApplied) {
      setVehicles([]);
      fetchVehicles();
    }
  }, [filterApplied, fetchVehicles]); // Add fetchVehicles as a dependency

  return {
    loading,
    vehicles,
    models,
    categories,
    fuelTypes,
    climateTypes,
    interiorColors,
    colorFilter,
    hasMore,
    error,
    filterApplied,

    resetFilters,
    loadMoreVehicles,
    handleModelChange,
    handleCategoryChange,
    handleFuelTypeChange,
    handleClimateChange,
    handleMinPriceChange,
    handleMaxPriceChange,
    handleMinRegYearChange,
    handleMaxRegYearChange,
    handleMinKmChange,
    handleMaxKmChange,
    handleMinEnginePowerChange,
    handleMaxEnginePowerChange,
    handleInteriorColorChange,
  };
};

export default useFetchVehicles;
