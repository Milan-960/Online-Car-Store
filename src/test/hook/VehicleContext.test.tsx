import React, { useContext } from "react";
import { render, screen } from "@testing-library/react";
import { VehicleProvider, VehicleContext } from "../../hooks/VehicleContext";

describe("VehicleProvider", () => {
  // The beforeAll function runs before all tests in this suite
  beforeAll(() => {
    // We mock window.scrollTo to avoid errors
    window.scrollTo = jest.fn();
  });

  // We define a test case "matches snapshot"
  it("matches snapshot", () => {
    // We use the render function to render the VehicleProvider component with a test div
    const { asFragment } = render(
      <VehicleProvider>
        <div>Test</div>
      </VehicleProvider>
    );

    // We check if the snapshot of this component matches the previous one
    expect(asFragment()).toMatchSnapshot();
  });

  // Define another test case "provides context values"
  it("provides context values", () => {
    // define a TestComponent which uses our context
    const TestComponent = () => {
      // the useContext function to get the current context
      const context = useContext(VehicleContext);

      // TestComponent renders some values from the context
      return (
        <div>
          {context.vehicles.length}
          {context.scrollPosition}
          {context.modelFilter}
          {context.categoryFilter}
        </div>
      );
    };

    // Render the VehicleProvider with the TestComponent as child
    render(
      <VehicleProvider>
        <TestComponent />
      </VehicleProvider>
    );

    // check if the values provided by the context are rendered correctly
    expect(screen.getByText("00")).toBeInTheDocument();
  });
});
