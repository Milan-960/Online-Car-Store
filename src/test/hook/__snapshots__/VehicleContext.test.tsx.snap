// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleProvider matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Test
  </div>
</DocumentFragment>
`;
